### PR TITLE
Simplify minor code around ThingTemplate class

### PR DESCRIPTION
**Pull request by xezon**
My review was too late on the last submission from @jonwil, so I put changes here.

### CODE REVIEW

File: src/game/common/rts/productionprerequisite.cpp [hunk 1]
```diff
@@ -26,24 +26,25 @@ void ProductionPrerequisite::Add_Unit_Prereq(Utf8String name, bool or_with_previ
 
 void ProductionPrerequisite::Resolve_Names()
 {
-    for (unsigned int i = 0; i < m_prereqUnit.size(); i++) {
```
Comment (xezon) [inline src/game/common/rts/productionprerequisite.cpp:29]: No need to call `m_prereqUnit.size()` on each iteration.

File: src/game/common/rts/productionprerequisite.cpp [hunk 2]
```diff
@@ -26,24 +26,25 @@ void ProductionPrerequisite::Add_Unit_Prereq(Utf8String name, bool or_with_previ
...
     }
 }
 
 int ProductionPrerequisite::Get_All_Possible_Build_Facility_Templates(ThingTemplate **tmpls, int max_tmpls)
 {
-    int count = 0;
+    int i = 0;
+    const int size = static_cast<int>(m_prereqUnit.size());
 
-    for (unsigned int i = 0; i < m_prereqUnit.size() && (i <= 0 || (m_prereqUnit[i].flags & 1) != 0) && count < max_tmpls;
```
Comment (xezon) [inline src/game/common/rts/productionprerequisite.cpp:43]: This function had 2 counters doing the same thing. Reduced to one counter and caching the size to simplify algorithm.

Comment (duncanspumpkin) [reply src/game/common/rts/productionprerequisite.cpp:43]: This is still not great. Is there no way to simplify this into something else.

File: src/game/common/thing/thingtemplate.cpp [hunk 3]
```diff
@@ -493,13 +493,13 @@ void ThingTemplate::Set_Copied_From_Default()
...
 void ThingTemplate::Resolve_Names()
 {
-    for (unsigned int i = 0; i < m_prerequisites.size(); i++) {
-        m_prerequisites[i].Resolve_Names();
+    for (ProductionPrerequisite &rerequisite : m_prerequisites) {
+        rerequisite.Resolve_Names();
     }
 
-    for (unsigned int i = 0; i < m_prerequisites.size(); i++) {
+    for (ProductionPrerequisite &rerequisite : m_prerequisites) {
```
Comment (duncanspumpkin) [inline src/game/common/thing/thingtemplate.cpp]: this goes for the other loop in this function as well `**p**rerequisite`

File: src/game/common/rts/productionprerequisite.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/game/common/rts/productionprerequisite.cpp:43]: I don't think you can change it like this count and i wont align at all times.

Comment (duncanspumpkin) [reply src/game/common/rts/productionprerequisite.cpp:43]: What about something like. I think this follows the original code:
```cpp
auto i = 0;
for (auto& prereq_unit : m_prereqUnit){
    if (i > 0 && ((prereq_unit.flags & 1) == 0)){
        break;
    }
    if (i > max_tmpls){
        break;
    }
    tmpls[i++] = prereq_unit.unit;
}
return i;
```

Comment (xezon) [reply src/game/common/rts/productionprerequisite.cpp:43]: It also has 2 iterators but ok. I like the splitting of the long condition. `i > max_tmpls` should be `i >= max_tmpls` I think.

File: src/game/common/thing/thingtemplate.cpp (hunk 3 above)
Comment (xezon) [reply src/game/common/thing/thingtemplate.cpp]: Ok.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**